### DISCUSSION TITLE
How to load chrome extensions?

### DISCUSSION THREAD

**Discussion by fanantoxa** [rubycdp/cuprite#208] [General]
Hey! I've found problem that I can't resolve by myself. Could you take a look please?

There are extensions in google chrome. Sometime in tests it's very useful to have some like react/redux devtools (or other) to debug bug in application.
Currently when I run test with cuprite (not heedlessly) extensions setting is disabled (so I can't install any).
Also with `Selenium::WebDriver::Chrome` I'm able to load `.crx` (file that can be downloaded from [store](https://chrome.google.com/webstore/category/extensions?hl=en))
Like:
```
Selenium::WebDriver::Chrome::Options.new.tap do |options|
  options.add_extension(<extension_path>)
end
```
and pass it to the `Capybara::Selenium::Driver` as options.

Is there any way how I can load Chrome with `.crx` extensions using cuprite driver?

**Comment by route** [comment]
Hi! I think the easiest thing to do is to use this [flag](https://peter.sh/experiments/chromium-command-line-switches/#load-extension) while running Chrome and you can pass them to the browser using https://github.com/rubycdp/ferrum#customization and `:browser_options` hash.

**Comment by tetienne** [comment]
@fanantoxa Did you succeed to load extension finally?

**Reply by route** [reply]
@tetienne what's the issue for you?

**Reply by tetienne** [reply]
When trying to load some chrome extensions doing:

```
options[:browser_options]["load-extension"] = [
  Rails.root.join("spec/e2e/config/extensions/vue-devtools.crx"),
  Rails.root.join("spec/e2e/config/extensions/apollo-devtools.crx"),
]
options[:browser_options]["disable-extensions-except"] = [
  Rails.root.join("spec/e2e/config/extensions/vue-devtools.crx"),
  Rails.root.join("spec/e2e/config/extensions/apollo-devtools.crx"),
]

Capybara.register_driver(:better_cuprite) do |app|
  Capybara::Cuprite::Driver.new(app, options)
end
```

I got this message when launching the test:

> Échec du chargement de l'extension depuis : . Fichier manifeste absent ou illisible

that I can translate to

> Fail to load the extension from: . Manifest file missing or not readable.

The only part I’m missing when reading the playwright documentation is the background context and the user dir. But I don’t know how to implement this with cuprite/ferrum.

**Reply by route** [reply]
Do you know that extensions don't work in headless Chrome?
Also this part looks fishy `Fail to load the extension from: .` what's this dot? something is wrong with paths

**Reply by tetienne** [reply]
I’ve well disabled the headless mode, the browser is visible. 
About the . I have no idea where it came from. `Rails.root.join("spec/e2e/config/extensions/vue-devtools.crx")` returns the expected path.

**Reply by route** [reply]
Check the flags of the running Chrome with `ps auxww` does it have needed paths?

**Comment by caifara** [comment]
Got a bit further using a string instead of an array. Needs the `disable` part as well.

``` ruby
        browser_options: {
          "load-extension" => react_devtools_path,
          "disable-extensions-except" => react_devtools_path
        }
```

The next problem is that the test runs in incognito mode and the extension isn't allowed in incognito mode.

**Reply by route** [reply]
That's true, but we are limited here by Chrome unfortunately
